<a href="https://colab.research.google.com/github/galyalina/Calendar/blob/master/Master_mmdetection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Training RepPoints on Toulouse dataset**

### **Overview**
This notebook install all the required libraries for Reppoint operation

### **Our Implementation**
Folked from https://github.com/microsoft/RepPoints

### **Our Data**

Source: http://rs.ipb.uni-bonn.de/data/semcity-toulouse/

Preprocessing steps:

1.   Crop images to 460px square images
2.   Generate bboxes from building segmantation masks
3.   Add to coco format json
4.   Dataset storage: 

### **Our Model**

TODO describe

#Mount dataset

In [1]:
# mount Google Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
%ls -l /usr/local | grep cuda

lrwxrwxrwx 1 root root   22 Feb  1 14:25 cuda -> /etc/alternatives/cuda/
drwxr-xr-x 1 root root 4096 Feb  1 14:17 cuda-10.0/
drwxr-xr-x 1 root root 4096 Feb  1 14:20 cuda-10.1/
lrwxrwxrwx 1 root root   25 Feb  1 14:25 cuda-11 -> /etc/alternatives/cuda-11/
drwxr-xr-x 1 root root 4096 Feb  1 14:22 cuda-11.0/
drwxr-xr-x 1 root root 4096 Feb  1 14:24 cuda-11.1/


In [3]:
# Check nvcc version
!nvcc -V
# Check GCC version
!gcc --version

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2020 NVIDIA Corporation
Built on Mon_Oct_12_20:09:46_PDT_2020
Cuda compilation tools, release 11.1, V11.1.105
Build cuda_11.1.TC455_06.29190527_0
gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.



# Install dependencies

In [4]:
# Get code
!rm -rf mmdetection
!git clone https://github.com/galyalina/mmdetection.git
%cd mmdetection

Cloning into 'mmdetection'...
remote: Enumerating objects: 23053, done.
remote: Counting objects: 100% (40/40), done.
remote: Compressing objects: 100% (30/30), done.
remote: Total 23053 (delta 12), reused 37 (delta 10), pack-reused 23013
Receiving objects: 100% (23053/23053), 25.95 MiB | 13.98 MiB/s, done.
Resolving deltas: 100% (16178/16178), done.
/content/mmdetection


In [5]:
# install dependencies: (use cu101 because colab has CUDA 10.1)
# !pip install -U torch==1.5.1+cu101 torchvision==0.6.1+cu101 -f https://download.pytorch.org/whl/torch_stable.html

# install mmcv-full thus we could use CUDA operators
!pip install mmcv-full==1.4.0 -f https://download.openmmlab.com/mmcv/dist/cu111/torch1.10.0/index.html

# Install mmdetection
!pip install -e .
!pip install -r requirements/lydorn_utils.txt
!pip install libs/lydorn_utils-0.0.1-py3-none-any.whl

Looking in links: https://download.openmmlab.com/mmcv/dist/cu111/torch1.10.0/index.html
     |████████████████████████████████| 58.0 MB 16.3 MB/s 
     |████████████████████████████████| 190 kB 8.4 MB/s 
Obtaining file:///content/mmdetection
  Running setup.py develop for mmdet
     |████████████████████████████████| 53 kB 2.3 MB/s 
     |████████████████████████████████| 97 kB 6.2 MB/s 
     |████████████████████████████████| 15.4 MB 58.7 MB/s 
     |████████████████████████████████| 880 kB 59.0 MB/s 
     |████████████████████████████████| 3.3 MB 59.1 MB/s 
     |████████████████████████████████| 11.2 MB 36.5 MB/s 
     |████████████████████████████████| 241 kB 79.7 MB/s 
     |████████████████████████████████| 63 kB 2.5 MB/s 
     |████████████████████████████████| 15.7 MB 30.7 MB/s 
     |████████████████████████████████| 47 kB 5.5 MB/s 
     |████████████████████████████████| 3.1 MB 71.2 MB/s 
     |████████████████████████████████| 97 kB 9.3 MB/s 
     |██████████████████████████

Processing ./libs/lydorn_utils-0.0.1-py3-none-any.whl


#!Restart Runtime!

# Alternative: install Anaconda env

In [ ]:
# %%bash
# git clone https://github.com/galyalina/mmdetection.git
# cd mmdetection
# ls
# bash setup-colab.sh

In [ ]:
# %%bash
# python setup.py develop

#Check installations

In [6]:
# Check Pytorch installation
import torch, torchvision
print(torch.__version__, torch.cuda.is_available())

# Check MMDetection installation
import mmdet
print(mmdet.__version__)

# Check mmcv installation
from mmcv.ops import get_compiling_cuda_version, get_compiler_version
print(get_compiling_cuda_version())
print(get_compiler_version())

1.10.0+cu111 True
2.21.0
11.1
GCC 7.3


#Link data folders from Drive

In [7]:
!ln -s "/content/drive/My Drive/MasterThesis/data" "/content/mmdetection/data"
!ln -sv "/content/drive/My Drive/MasterThesis/work_dirs" "/content/mmdetection/work_dirs"
!ln -sv "/content/drive/My Drive/MasterThesis/experiments" "/content/mmdetection/experiments"

'/content/mmdetection/work_dirs' -> '/content/drive/My Drive/MasterThesis/work_dirs'
'/content/mmdetection/experiments' -> '/content/drive/My Drive/MasterThesis/experiments'


#Visualize network



In [ ]:
!pip install torchsummary

#Add visualization

In [ ]:
!git clone https://github.com/Stephenfang51/mmdetection_visualize

/bin/bash: line 0: cd: /root/content: No such file or directory
Cloning into 'mmdetection_visualize'...
remote: Enumerating objects: 53, done.
remote: Total 53 (delta 0), reused 0 (delta 0), pack-reused 53
Unpacking objects: 100% (53/53), done.


In [ ]:
!cp mmdetection_visualize/voc_eval_visualize.py tools/
!cp mmdetection_visualize/mean_ap_visualize.py mmdet/core/evaluation/

In [ ]:
!cp work_dirs/reppoints_moment_r101_fpn_gn-neck+head_2x_toulouse/20211125_095409.log.json mmdetection_visualize/json/

cp: cannot create regular file 'mmdetection_visualize/json/': No such file or directory


In [ ]:
%%bash 
cd mmdetection/mmdetection_visualize
python visualize.py json/20211125_095409.log.json

# Evaluate zeven dataset

In [ ]:
%%bash      
python tools/test.py \
    work_dirs/reppoints_moment_r101_fpn_gn-neck+head_2x_toulouse/test_reppoints_moment_r101_fpn_gn-neck+head_2x_toulouse.py \
    work_dirs/reppoints_moment_r101_fpn_gn-neck+head_2x_toulouse/epoch_24.pth \
    --eval bbox\
    --show-dir data/brauschweig/results \

#Zeven labled result

loading annotations into memory...
Done (t=0.01s)
creating index...
index created!
load checkpoint from local path: work_dirs/reppoints_moment_r101_fpn_gn-neck+head_2x_toulouse/epoch_24.pth
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 112/112, 11.1 task/s, elapsed: 10s, ETA:     0s
Evaluating bbox...
Loading and preparing results...
DONE (t=0.01s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=8.90s).
Accumulating evaluation results...
DONE (t=0.24s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.430
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.726
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.452
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.361
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.550
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.445
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.529
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.529
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.529
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.469
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.639
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.629

#Train model, uncomment for training

In [8]:
%%bash
# source activate openmmlab

#Add resume_from = 'work_dirs/reppoints_moment_r101_fpn_gn-neck+head_2x_toulouse/epoch_23.pth' for resuming training
#RepPoints
bash tools/dist_train.sh configs/reppoints_toulouse/reppoints_moment_r101_fpn_gn-neck+head_2x_toulouse.py 1 &>> work_dirs/reppoints_moment_r101_fpn_gn-neck+head_2x_toulouse/log.txt
#RepPoints
# bash tools/dist_train.sh configs/reppoints/reppoints_moment_r101_fpn_dconv_c3-c5_gn-neck+head_2x_toulouse.py 1 &>> work_dirs/reppoints_moment_r101_fpn_dconv_c3-c5_gn-neck+head_2x_toulouse/log.txt
#YOLO
# bash tools/dist_train.sh configs/yolo/yolov3_d53_320_273e_toulouse.py 1 &>> work_dirs/yolov3_d53_320_273e_toulouse/log.txt

#Faster RCNN
# bash tools/dist_train.sh configs/faster_rcnn_toulouse/faster_rcnn_r101_fpn_2x_toulouse.py 1 &>> work_dirs/faster_rcnn_r101_fpn_2x_toulouse/log.txt

In [ ]:
%load_ext tensorboard --logdir faster_rcnn_toulouse/tf_logs

#Print predictions

In [ ]:
#Required for slicing the image 
!pip install sahi

In [ ]:
%%bash 
python tools/analysis_tools/get_flops.py 'work_dirs/reppoints_v1_r101_fpn_giou_mstrain_2x_toulouse/reppoints_v1_r101_fpn_giou_mstrain_2x_toulouse.py'
# python tools/analysis_tools/get_flops.py 'work_dirs/reppoints_moment_r101_fpn_gn-neck+head_2x_toulouse/reppoints_moment_r101_fpn_gn-neck+head_2x_toulouse.py' 

In [ ]:
from mmdet.apis import inference_detector, init_detector
from mmdet.apis import show_result_pyplot
import os
import mmcv
import numpy as np
import lydorn_utils.geo_utils as utils
import numpy as np
import rasterio
from PIL import Image, ImageDraw
from sahi.model import MmdetDetectionModel
from sahi.predict import get_prediction, get_sliced_prediction, predict
from sahi.utils.cv import visualize_object_predictions

# Choose to use a config and initialize the detector
config = 'mmdetection/work_dirs/reppoints_moment_r101_fpn_gn-neck+head_2x_toulouse/reppoints_moment_r101_fpn_gn-neck+head_2x_toulouse.py'
# Setup a checkpoint file to load
checkpoint = 'mmdetection/work_dirs/reppoints_moment_r101_fpn_gn-neck+head_2x_toulouse/checkpoint_24.pth'
# initialize the detector
model = init_detector(config, checkpoint, device='cuda:0')
detection_model = MmdetDetectionModel(
    model_path=checkpoint,
    config_path=config,
    confidence_threshold=0.4,
    device='cuda' # or 'cpu'
  )

# test_directory = 'mmdetection/data/niedersachsen/brauschwieg/rgb_reduced/'
# prediction_directory = 'mmdetection/data/niedersachsen/brauschwieg/predictions/sliced_460_px/'
# osm_and_prediction_directory = 'mmdetection/data/niedersachsen/brauschwieg/osm_and_predictions/'

test_directory = 'mmdetection/data/niedersachsen/zeven/rgb_reduced/'
prediction_directory = 'mmdetection/data/niedersachsen/zeven/predictions/sliced_460/'
osm_and_prediction_directory = 'mmdetection/data/niedersachsen/zeven/osm_and_predictions/'

# test_directory = 'mmdetection/data/toulouse/unlabled/'
# prediction_directory = 'mmdetection/data/toulouse/predictions/sliced_460_px/'
# osm_and_prediction_directory = 'mmdetection/data/toulouse/osm_and_predictions/'

def get_osm_segmentation(original_dir, predicted_dir, image_name, destination_dir):
    mask = Image.open(predicted_dir + image_name)
    print(predicted_dir + image_name)
    mask_draw = ImageDraw.Draw(mask)
    str_proj_EPSG_25832 = "PROJCS[\"ETRS89 / UTM zone 32N\",GEOGCS[\"ETRS89\",DATUM[\"European_Terrestrial_Reference_System_1989\",SPHEROID[\"GRS 1980\",6378137,298.257222101,AUTHORITY[\"EPSG\",\"7019\"]],TOWGS84[0,0,0,0,0,0,0],AUTHORITY[\"EPSG\",\"6258\"]],PRIMEM[\"Greenwich\",0,AUTHORITY[\"EPSG\",\"8901\"]],UNIT[\"degree\",0.0174532925199433,AUTHORITY[\"EPSG\",\"9122\"]],AUTHORITY[\"EPSG\",\"4258\"]],PROJECTION[\"Transverse_Mercator\"],PARAMETER[\"latitude_of_origin\",0],PARAMETER[\"central_meridian\",9],PARAMETER[\"scale_factor\",0.9996],PARAMETER[\"false_easting\",500000],PARAMETER[\"false_northing\",0],UNIT[\"metre\",1,AUTHORITY[\"EPSG\",\"9001\"]],AXIS[\"Easting\",EAST],AXIS[\"Northing\",NORTH],AUTHORITY[\"EPSG\",\"25832\"]]"
    poligons = utils.get_polygons_from_osm(original_dir + image_name,
                                           tag="building",
                                           ij_coords=False,
                                           specific_projection=str_proj_EPSG_25832)
    for p in poligons:
      result = list(map(tuple, np.array(p).astype(int)))
      mask_draw.polygon(result, outline='#ff00ff')
      bbox = p[:, 0].min(), p[:, 0].max(), p[:, 1].min(), p[:, 1].max()
      mask_draw.rectangle([(bbox[0], bbox[2]), (bbox[1], bbox[3])], outline='#ff00ff', width=2)
    mask.save(destination_dir + image_name)
    return destination_dir + image_name


def store_predictions_to_file(original_image, results, output):
    model.show_result(original_image,
                    results,
                    score_thr=0.4,
                    bbox_color=(00, 255, 00),
                    text_color=(00, 255, 00),
                    mask_color=None,
                    thickness=2,
                    font_size=1,
                    win_name='',
                    show=False,
                    wait_time=0,
                    out_file=output)

for filename in os.listdir(test_directory):
    if (filename.endswith(".tif") or filename.endswith(".tiff") or filename.endswith(".jpeg") or filename.endswith(".png")) and '_detection' not in filename: 
        # result = inference_detector(model, test_directory+filename)
        # result_sliced = get_sliced_prediction(
        #       test_directory+filename,
        #       detection_model, 
        #       slice_width=460, 
        #       slice_height=460,
        #       overlap_height_ratio = 0,
        #       overlap_width_ratio = 0
        # )
        # result = detection_model._create_original_predictions_from_object_prediction_list(result_sliced.object_prediction_list)
        # store_predictions_to_file(test_directory+filename, result, prediction_directory+filename)

        osm_mask_path = get_osm_segmentation(test_directory, prediction_directory, filename, osm_and_prediction_directory)
        # store_predictions_to_file(osm_and_prediction_directory+filename, result, osm_and_prediction_directory+filename)
        # break
    else:
        continue

Test trained model

In [ ]:
while True:pass

PYTORCH TO ONNX (EXPERIMENTAL)

In [ ]:
python tools/deployment/pytorch2onnx.py \
    configs/yolo/yolov3_d53_mstrain-608_273e_coco.py \
    checkpoints/yolo/yolov3_d53_mstrain-608_273e_coco.pth \
    --output-file checkpoints/yolo/yolov3_d53_mstrain-608_273e_coco.onnx \
    --input-img demo/demo.jpg \
    --test-img tests/data/color.jpg \
    --shape 608 608 \
    --show \
    --verify \
    --dynamic-export \
    --cfg-options \
      model.test_cfg.deploy_nms_pre=-1 \